File: problems/Word Break II/wordBreak.cs
public class Solution {
    public IList<string> WordBreak(string s, IList<string> wordDict) {
        IList<string>[] cache = new IList<string>[s.Length];
        ISet<string> set = wordDict.ToHashSet();
        var res = Helper(ref s, 0, cache, set);
        return res;
    }

    private IList<string> Helper(ref string s, int idx, IList<string>[] cache, ISet<string> set)
    {
        if (cache[idx] != null)
        {
            return cache[idx];
        }

        IList<string> res = new List<string>();

        for (int l = 1; l <= s.Length - idx; l++)
        {
            var word = s.Substring(idx, l);
            if (set.Contains(word))
            {
                var next = idx + l;
                if (next == s.Length)
                {
                    res.Add(word);
                }
                else
                {
                    var inner = Helper(ref s, next, cache, set);
                    foreach (var subsentence in inner)
                    {
                        res.Add($"{word} {subsentence}");
                    }
                }
            }
        }

        cache[idx] = res;
        return res;
    }
}

File: problems/Word Break II/wordBreak.java
class Solution {
    public List<String> wordBreak(String s, List<String> wordDict) {
        List<String> result = new ArrayList<>();

        dfs(s, wordDict, new ArrayList<String>(), result);

        return result;
    }

    private void dfs(String text, List<String> wordDict, List<String> path, List<String> result) {
        if ((null == text || 0 == text.length()) && 0 < path.size()) {
            result.add(String.join(" ", path));
            return;
        }

        for (String word : wordDict) {
            if (checkPrefix(text, word)) {
                List<String> newPath = new ArrayList<>(path);

                newPath.add(word);
                dfs(text.substring(word.length()), wordDict, newPath, result);
            }
        }
    }

    private boolean checkPrefix(String text, String pattern) {
        if (text.length() < pattern.length()) {
            return false;
        }

        for (int idx = 0; pattern.length() > idx; ++idx) {
            if (text.charAt(idx) != pattern.charAt(idx)) {
                return false;
            }
        }

        return true;
    }
}
